boskos/boskos.go

```diff
@@ -0,0 +1,275 @@
+/*
+Copyright 2017 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package main
+
+import (
+	"encoding/json"
+	"flag"
+	"fmt"
+	"io/ioutil"
+	"net/http"
+	"os"
+	"sync"
+	"time"
+
+	"github.com/Sirupsen/logrus"
+)
+
+var configPath = flag.String("config", "projects.json", "Path to init projects file")
+var storage = flag.String("storage", "", "Path to presistent volume to load the state")
+
+func main() {
+	flag.Parse()
+	logrus.SetFormatter(&logrus.JSONFormatter{})
+
+	r, err := NewRanch(*configPath, *storage)
+	if err != nil {
+		logrus.WithError(err).Fatal("Fail to create my ranch!")
+	}
+
+	http.Handle("/", handleDefault())
+	http.Handle("/request", handleRequestProject(r))
```

Comment:
*Atomic find and Move*:
* Find any project in state S1
* Move P into state S2 for duration D
* Concrete example: Find a dirty project and move it into state cleaning for 10m
* Concrete result: return project k8s-jkns-whatever at timestamp 123

*Atomic move specific*:
* Move project P from S1 to S2 for duration D
* Concrete example: move k8s-blah-blah at timestamp 555 from state busy to state hospitalized
* Concrete result: Ignored, k8s-blah-blah is no longer at timestamp 555

Additional ideas:

*List*:
* Find the number of projects in state S1
* Concrete example: find all projects in a hospitalized state
* Concrete result: [P1@223, P2@224, P3@111, ...]

Hmm... Ideally you could replace list with a more complex find and move... but that seems complicated